guava/src/com/google/common/util/concurrent/Uninterruptibles.java

```diff
@@ -326,7 +326,67 @@ public static boolean tryAcquireUninterruptibly(
     }
   }
 
-  // TODO(user): Add support for waitUninterruptibly.
+  /**
+   * Invokes {@code object.}{@link Object#wait() wait()} uninterruptibly.
+   *
+   * @param object to wait upon.
+   *
+   * @throws IllegalMonitorStateException if the current thread is not
+   *               the owner of the object's monitor.
+   */
+  @GwtIncompatible // concurrency
+  public static void waitUninterruptibly(Object object) {
+    Preconditions.checkNotNull(object);
+    boolean interrupted = false;
+    try {
+      while (true) {
+        try {
+          object.wait();
+          return;
+        } catch (InterruptedException e) {
+          interrupted = true;
```

Comment:
Hmm. I am pretty sure that we need to `return` here.

`wait` is different than the other interruptible methods because `notify` is a one-shot thing: If you aren't `wait`-ing during the call, then you don't get the notification. This is different than something like `latch.await()`, where you can start waiting after the `countDown()` call and still "receive the notification."

Past discussion converged on the idea that we'd probably need to do something like this:

```
public static void waitUninterruptibly(Object waitOn) {
  boolean reinterrupt = Thread.interrupted();
  try {
    waitOn.wait();
  } catch (InterruptedException e) {
    reinterrupt = true;
  }
  if (reinterrupt) {
    selfInterrupt();
  }
}
```

Sorry for not digging this up earlier. I'd forgotten about it until I actually saw the code :(